[wip][LLD] Support RISCV vendor-specific relocations.

resistor opened the pull request:
This is an implentation of RISCV vendor-specific relocation support in LLD, derived from a working prototype implementation in the downstream CHERIoT repository: https://github.com/CHERIoT-Platform/llvm-project/commit/3d6d6f7d9480b590731cbcf4b4817e1fa3049854
At this time I am seeking design feedback on the shape of the implementation, while I work on developing tests that do not depend on CHERIoT.

Key elements of the design include:
 - Representing vendor-namespaced relocs by treating RelType as a bitfield
 - Creating a reloc iterator that can fold R_RISCV_VENDOR relocs into to the bitfield on the following reloc.

lenary commented on lld/ELF/Arch/RISCVInternalRelocations.h:
```diff
@@ -0,0 +1,113 @@
+// Bits 9 -> 31 of RelType are used to indicate vendor-specific relocations.
+constexpr uint32_t INTERNAL_RISCV_VENDOR_MASK = 0xFFFFFFFF << 9;
+constexpr uint32_t INTERNAL_RISCV_VENDOR_QUALCOMM = 1 << 9;
+constexpr uint32_t INTERNAL_RISCV_VENDOR_ANDES = 1 << 10;
```
Why does this need to be a mask? No relocation should have both bits set, right?

Surely we can effectively do this by viewing the 32-bit field as a 8 bit field (at the lsbs) and a 24-bit field (the msbs), where the 24-bit field is 0 for standard relocations, 1 for internal relocations, and 2+ for vendor relocations. I don't think we'll need to be more efficient than that, even though each vendor can only have 64 relocations, not 256.

I tried something more complex in ELD, which in retrospect was too clever.

```suggestion
constexpr uint32_t INTERNAL_RISCV_MASK = 0xFFFFFF << 8;
constexpr uint32_t INTERNAL_RISCV_INTERNAL = 1 << 8;
constexpr uint32_t INTERNAL_RISCV_VENDOR_QUALCOMM = 2 << 8;
constexpr uint32_t INTERNAL_RISCV_VENDOR_ANDES = 3 << 8;

constexpr uint32_t INTERNAL_R_RISCV_GPREL_I = INTERNAL_RISCV_INTERNAL | 0;
constexpr uint32_t INTERNAL_R_RISCV_GPREL_S = INTERNAL_RISCV_INTERNAL | 1;
constexpr uint32_t INTERNAL_R_RISCV_X0REL_I = INTERNAL_RISCV_INTERNAL | 2;
constexpr uint32_t INTERNAL_R_RISCV_X0REL_S = INTERNAL_RISCV_INTERNAL | 3;
```

lenary commented on lld/ELF/Target.cpp:
```diff
@@ -40,6 +41,12 @@ using namespace lld::elf;
+    if (VendorString)
+      s = getRISCVVendorRelocationTypeName(type & ~INTERNAL_RISCV_VENDOR_MASK,
+                                           *VendorString);
+  }
```
Do we want to do a fallback here for `R_RISCV_CUSTOM<n>` or not? Presumably we only get here when LLD managed some but not enough processing of the reloc, so maybe this is only going to be called for relocation numbers lld knows about?

resistor replied:
I'll add the fallback path. LLD should print *something* sane if run on object files containing vendor-relics it doesn't (yet) understand. If we don't include the relocation number, it makes it impossible for user to figure out what's going wrong.

lenary commented on lld/ELF/SyntheticSections.cpp:
```diff
@@ -1703,6 +1704,18 @@ void DynamicReloc::finalize(Ctx &ctx, SymbolTableBaseSection *symt) {
+    for (const auto &reloc : relocs) {
+      if (reloc.type.v & INTERNAL_RISCV_VENDOR_MASK) {
+        size += entsize;
+      }
```
Alright, I see why you split bit 8 and bit 9, so this will need a slight update with my scheme. I still think mine is slightly neater, even with this check.

resistor replied:
I ended up opting not to merge the internal bit with the others, as this is _not_ the only spot that would need updating. It could still be done if you feel strongly about it.

lenary commented on lld/ELF/SyntheticSections.cpp:
```diff
@@ -1725,6 +1738,47 @@ void RelocationBaseSection::computeRels() {
   }
+  // Insert R_RISCV_VENDOR relocations very late, so that it doesn't interfere
+  // with relocation sorting above.
+  if (ctx.arg.emachine == EM_RISCV) {
```
Neat

lenary reviewed (commented):
I like the iterator, it seems clean and a sensible wrapper for the logic.

I have more feedback about the bitfields, at the moment, but overall I'm happy with the direction so far.

I haven't yet had to think about dynamic relocations, and if they have some differences compared to the existing design/spec (are local symbols a problem here?)

lenary commented on lld/ELF/SyntheticSections.cpp:
```diff
@@ -1725,6 +1738,47 @@ void RelocationBaseSection::computeRels() {
+  auto riscvVendorString = getRISCVVendorString(reloc.type);
+  if (ctx.arg.emachine == llvm::ELF::EM_RISCV && riscvVendorString) {
+    Symbol &vendorSym = *ctx.symtab->addSymbol(Defined{
+        ctx, ctx.internalFile, *riscvVendorString, llvm::ELF::STB_GLOBAL,
```
Oracle docs seem to agree with you:

> The .dynsym table begins with the standard NULL symbol, followed by the files global symbols. STT_FILE symbols are typically not present in this symbol table. STT_SECTION symbols might be present if required by relocation entries.

From: https://docs.oracle.com/cd/E23824_01/html/819-0690/chapter6-79797.html#stlac

I will go back to the ABI about this. We made them local to echo mapping symbols, but evidently that doesn't work for vendor-specific dynamic relocations.

lenary commented on lld/ELF/Arch/RISCV.cpp:
```diff
@@ -1533,3 +1540,21 @@ template <class ELFT> void RISCV::scanSection1(InputSectionBase &sec) {
 }
+
+namespace lld::elf {
+uint32_t getRISCVVendorRelType(StringRef rvVendor) {
```
Nit: This could maybe be something like the following for clarity, given it now doesn't return a type:

```suggestion
uint32_t getRISCVVendorRelMarker(StringRef rvVendor) {
```

lenary reviewed (commented):
This seems clear to me what's going on (though I've implemented this once already). I would be happy to help review/maintain this code if it was landed.

One Nit below.

Do we want to make sure we never parse a relocation from a file that's more than 255? They would have had to have been made deliberately, but are possible to encode on rv64. It's the sort of error that's useful to have and remove in the future, if the abi ever gets so many relocations.

resistor committed:
[LLD] Add support for statically resolved vendor-specific RISCV relocations.

This is achieved by using some of the bits of RelType to tag vendor namespaces.
This change also adds a relocation iterator for RISCV that folds vendor namespaces
into the RelType of the following relocation.

This patch is extracted from the implementation of RISCV vendor-specific relocations
in the CHERIoT LLVM downstream: https://github.com/CHERIoT-Platform/llvm-project/commit/3d6d6f7d9480b590731cbcf4b4817e1fa3049854

resistor commented:
Superseded by https://github.com/llvm/llvm-project/pull/169273 (static relocation support) and https://github.com/llvm/llvm-project/pull/169274 (dynamic relocation support).
